flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/DDLOperationConverter.java

```diff
@@ -0,0 +1,808 @@
+        Map<String, String> props = new HashMap<>(oldTable.getOptions());
+        props.put(ALTER_TABLE_OP, ALTER_COLUMNS.name());
+        if (cascade) {
+            props.put(ALTER_COL_CASCADE, "true");
+        }
+        return new AlterTableSchemaOperation(
+                tableIdentifier,
+                new CatalogTableImpl(
+                        newSchema, oldTable.getPartitionKeys(), props, oldTable.getComment()));
+    }
+
+    private Operation convertAddReplaceColumns(
+            ObjectIdentifier tableIdentifier,
+            CatalogTable oldTable,
+            List<FieldSchema> newCols,
+            boolean replace,
+            boolean cascade) {
+        // prepare properties
+        Map<String, String> props = new HashMap<>(oldTable.getOptions());
+        props.put(ALTER_TABLE_OP, ALTER_COLUMNS.name());
+        if (cascade) {
+            props.put(ALTER_COL_CASCADE, "true");
+        }
+        TableSchema oldSchema = oldTable.getSchema();
+        final int numPartCol = oldTable.getPartitionKeys().size();
+        TableSchema.Builder builder = TableSchema.builder();
+        // add existing non-part col if we're not replacing
+        if (!replace) {
+            List<TableColumn> nonPartCols =
+                    oldSchema.getTableColumns().subList(0, oldSchema.getFieldCount() - numPartCol);
+            for (TableColumn column : nonPartCols) {
+                builder.add(column);
+            }
+            setWatermarkAndPK(builder, oldSchema);
+        }
+        // add new cols
+        for (FieldSchema col : newCols) {
+            builder.add(
+                    TableColumn.physical(
+                            col.getName(),
+                            HiveTypeUtil.toFlinkType(
+                                    TypeInfoUtils.getTypeInfoFromTypeString(col.getType()))));
+        }
+        // add part cols
+        List<TableColumn> partCols =
+                oldSchema
+                        .getTableColumns()
+                        .subList(oldSchema.getFieldCount() - numPartCol, oldSchema.getFieldCount());
+        for (TableColumn column : partCols) {
+            builder.add(column);
+        }
+        return new AlterTableSchemaOperation(
+                tableIdentifier,
+                new CatalogTableImpl(
+                        builder.build(),
+                        oldTable.getPartitionKeys(),
+                        props,
+                        oldTable.getComment()));
+    }
+
+    private Operation convertAlterTableProps(
+            ObjectIdentifier tableIdentifier,
+            CatalogTable oldTable,
+            CatalogPartitionSpec partSpec,
+            CatalogPartition catalogPartition,
+            Map<String, String> newProps) {
+        Map<String, String> props = new HashMap<>();
+        if (catalogPartition != null) {
+            props.putAll(catalogPartition.getProperties());
+            props.putAll(newProps);
```

Comment:
Why put `newProperties` into partition properties? Will the `newProperties` be table properites?

Reply:
Lots of hive's ALTER semantics are encoded as properties, e.g. changing locations, file format, etc. So we need the `CatalogPartition` to carry these properties.